File: otb/src/main/java/com/orange/essentials/otb/manager/TrustBadgeManager.java
/*
 * Copyright 2016 Orange
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* Orange Trust Badge library
 *
 * Module name: com.orange.essentials:otb
 * Version:     1.0
 * Created:     2016-03-15 by Aurore Penault, Vincent Boesch, and Giovanni Battista Accetta
 */
package com.orange.essentials.otb.manager;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import com.orange.essentials.otb.event.EventTagger;
import com.orange.essentials.otb.event.EventType;
import com.orange.essentials.otb.model.Term;
import com.orange.essentials.otb.model.TrustBadgeElement;
import com.orange.essentials.otb.model.type.ElementType;
import com.orange.essentials.otb.model.type.GroupType;
import com.orange.essentials.otb.model.type.RatingType;
import com.orange.essentials.otb.model.type.UserPermissionStatus;

import java.util.ArrayList;
import java.util.List;

/**
 * <p/>
 * File name:   TrustBadgeManager
 * Main singleton handling the trustbadge elements
 */
public enum TrustBadgeManager {
    INSTANCE;

    /**
     * Log TAG
     */
    private static final String TAG = "TrustBadgeManager";
    /**
     * mApplicationName
     * Host application name
     */
    private String mApplicationName;
    /**
     * mApplicationPackageName
     * Host application packageName
     */
    private String mApplicationPackageName;

    /**
     * Provides a way to tag events
     */
    private EventTagger mEventTagger;

    /**
     * Store all informations about Main Data, Complementary Data, and Usage Data
     */
    private List<TrustBadgeElement> mTrustBadgeElements = new ArrayList<>();

    /**
     * Stores all the listeners who should act when a toggable badge element value is switched
     */
    private List<TrustBadgeElementListener> mTrustBadgeElementListeners = new ArrayList<>();
    /**
     * Stores all the listeners who should act when a toggable badge element value is switched
     */
    private List<BadgeListener> mBadgeListeners = new ArrayList<>();

    /**
     * List of terms and conditions that should be displayed in "terms and conditions" section
     */
    private List<Term> mTerms = new ArrayList<>();
    /**
     * Boolean indicating if app checked improvement program
     */
    private boolean mUsingImprovementProgram = true;
    /**
     * boolean letting us know if badges are initialized, false by default
     */
    private boolean mInitialized = false;

    /** METHODS ********************************************************************************* */

    /** KEYS DATAS */

    /**
     * Return the host application name
     *
     * @return Host application name from AndroidManifest.xml
     */
    private String extractApplicationName(@NonNull Context context) {
        String appName = context.getPackageName();
        Log.d(TAG, "extractApplicationName");
        CharSequence csName = context.getApplicationInfo().loadLabel(context.getPackageManager());
        if (null != csName) {
            appName = csName.toString();
        }
        Log.d(TAG, "extractApplicationName");
        return appName;
    }

    /**
     * provide init with default BadgeFactory
     *
     * @param context : Context providing needed data to retrieve permissions infos
     */
    public void initialize(@NonNull Context context) {
        PermissionManager.INSTANCE.initPermissionList(context);
        initialize(context, TrustBadgeElementFactory.getDefaultElements(context), new ArrayList<Term>());
    }


    /**
     * initialize
     *
     * @param context            : Activity providing needed data to retrieve permissions infos
     * @param trustBadgeElements : the list of trsut badge that will be used into the app
     */
    public void initialize(@NonNull Context context, List<TrustBadgeElement> trustBadgeElements, List<Term> terms) {

        Log.d(TAG, "TrustBadgeManager initialize");
        PermissionManager.INSTANCE.initPermissionList(context);
        mApplicationName = extractApplicationName(context);
        if (!mApplicationName.isEmpty()) {
            Log.d(TAG, "mApplicationName is NOT empty");
            mApplicationPackageName = context.getPackageName();
            if (mApplicationPackageName != null && mApplicationPackageName.length() != 0) {
                Log.d(TAG, "mApplicationPackageName is NOT empty");
                /** Initialisation of TrustBadgeElements to 15 permissions */
                setTrustBadgeElements(trustBadgeElements);
                setTerms(terms);
            } else {
                Log.d(TAG, "Context does not provide PackageName.");
            }
        } else {
            Log.d(TAG, "Context does not provide ApplicationName.");
        }
        mInitialized = true;
    }

    /**
     * retrieve list of permissions about kids
     */
    public ArrayList<TrustBadgeElement> getElementsForUsage() {
        ArrayList<TrustBadgeElement> resultPermissionList = new ArrayList<>();
        for (int i = 0; i < mTrustBadgeElements.size(); i++) {
            TrustBadgeElement trustBadgeElement = mTrustBadgeElements.get(i);
            if (null != trustBadgeElement && trustBadgeElement.getElementType() == ElementType.USAGE) {
                resultPermissionList.add(trustBadgeElement);
            }
        }
        return resultPermissionList;
    }

    /**
     * retrieve list of  permissions
     */
    public ArrayList<TrustBadgeElement> getElementsForDataCollected() {
        Log.d(TAG, "getElementsForDataCollected");
        ArrayList<TrustBadgeElement> resultPermissionList = new ArrayList<>();
        if (mTrustBadgeElements != null) {
            for (TrustBadgeElement trustBadgeElement : mTrustBadgeElements) {
                Log.d(TAG, "getElementsForDataCollected  : " + trustBadgeElement.getGroupType() + ", " + trustBadgeElement.getElementType() + ", " + trustBadgeElement.getAppUsesPermission() + ", " + trustBadgeElement.getUserPermissionStatus());
                if (trustBadgeElement.getElementType() != ElementType.USAGE) {
                    resultPermissionList.add(trustBadgeElement);
                    Log.d(TAG, trustBadgeElement.getNameKey() + " ==> ADDED");
                }
            }
        } else {
            Log.d(TAG, "Empty mTrustBadgeElements.");
        }
        return resultPermissionList;
    }

    /**
     * retrieve a specific badge according to type
     */
    public TrustBadgeElement getSpecificPermission(GroupType groupType) {
        Log.d(TAG, "getSpecificPermission");
        TrustBadgeElement result = null;
        if (null != mTrustBadgeElements) {
            for (TrustBadgeElement trustBadgeElement : mTrustBadgeElements) {
                if (trustBadgeElement.getGroupType() == groupType) {
                    Log.d(TAG, "trustBadgeElement.getElementType() equals " + ElementType.USAGE.toString());
                    result = trustBadgeElement;
                }
            }
        }
        return result;
    }

    /**
     * GETTERS & SETTERS************************************************************************
     */
    /**
     * Gets the array of trustbadge elements
     *
     * @return the list of TrustBadgeElements that will be displayed to the user
     */
    public List<TrustBadgeElement> getTrustBadgeElements() {
        return mTrustBadgeElements;
    }

    public void setTrustBadgeElements(List<TrustBadgeElement> trustBadgeElements) {
        mTrustBadgeElements = trustBadgeElements;
    }

    public void addTrustBadgeElement(TrustBadgeElement trustBadgeElement) {
        if (null != trustBadgeElement) {
            if (null == mTrustBadgeElements) {
                mTrustBadgeElements = new ArrayList<>();
            }
            mTrustBadgeElements.add(trustBadgeElement);
        }
    }

    /**
     * Remove all element from the SparseArray and the array
     */
    public void clearTrustBadgeElements() {
        mTrustBadgeElements.clear();
    }


    /**
     * Refresh the badges according to the permissions granted by the app
     */
    public void refreshTrustBadgePermission(@NonNull Context context) {
        Log.d(TAG, "refreshTrustBadgePermission...");
        PermissionManager.INSTANCE.initPermissionList(context);
        for (int i = 0; i < mTrustBadgeElements.size(); i++) {
            TrustBadgeElement trustBadgeElement = mTrustBadgeElements.get(i);
            if (trustBadgeElement.isShouldBeAutoConfigured()) {
                GroupType type = trustBadgeElement.getGroupType();
                if (type.isSystemPermission()) {
                    trustBadgeElement.setUserPermissionStatus(PermissionManager.INSTANCE.doesUserAlreadyAcceptPermission(context, trustBadgeElement.getGroupType()));
                }
                if (type.equals(GroupType.IMPROVEMENT_PROGRAM)) {
                    trustBadgeElement.setUserPermissionStatus(mUsingImprovementProgram ? UserPermissionStatus.GRANTED : UserPermissionStatus.NOT_GRANTED);
                }
            }

            Log.d(TAG, "refresh = " + trustBadgeElement);
        }

    }

    public String getApplicationName() {
        return mApplicationName;
    }

    /**
     * getPegiAge : Retrieve permission pegi age rating
     *
     * @param trustBadgeElement : th badge to retrieve pegi age value text
     * @return String age permission
     */
    public String getPegiAge(TrustBadgeElement trustBadgeElement) {
        String ageInNumber = "";
        String ageRating = trustBadgeElement.getDescriptionKey();
        if (ageRating.contains(RatingType.THREE.name())) {
            ageInNumber = Integer.toString(RatingType.THREE.getAge());
        }
        if (ageRating.contains(RatingType.SEVEN.name())) {
            ageInNumber = Integer.toString(RatingType.SEVEN.getAge());
        }
        if (ageRating.contains(RatingType.TWELVE.name())) {
            ageInNumber = Integer.toString(RatingType.TWELVE.getAge());
        }
        if (ageRating.contains(RatingType.SIXTEEN.name())) {
            ageInNumber = Integer.toString(RatingType.SIXTEEN.getAge());
        }
        if (ageRating.contains(RatingType.EIGHTEEN.name())) {
            ageInNumber = Integer.toString(RatingType.EIGHTEEN.getAge());
        }
        return ageInNumber;
    }

    public String getApplicationPackageName() {
        return mApplicationPackageName;
    }

    public List<Term> getTerms() {
        return mTerms;
    }

    public void setTerms(List<Term> terms) {
        mTerms = terms;
    }

    public EventTagger getEventTagger() {
        if (null == mEventTagger) {
            //Provides a default eventtagger if none has been provided
            mEventTagger = new EventTagger() {
                @Override
                public void tag(EventType eventType) {
                    Log.d(TAG, "tagging event " + eventType);
                }

                @Override
                public void tagElement(EventType eventType, TrustBadgeElement element) {
                    Log.d(TAG, "tagging event " + eventType + " for element " + element);
                }
            };
        }
        return mEventTagger;
    }

    public void setEventTagger(EventTagger eventTagger) {
        mEventTagger = eventTagger;
    }

    public boolean isInitialized() {
        return mInitialized;
    }

    public boolean isUsingImprovementProgram() {
        return mUsingImprovementProgram;
    }

    public void setUsingImprovementProgram(boolean usingImprovementProgram) {
        mUsingImprovementProgram = usingImprovementProgram;
        TrustBadgeElement improvementProgram = TrustBadgeManager.INSTANCE.getSpecificPermission(GroupType.IMPROVEMENT_PROGRAM);
        if (null != improvementProgram) {
            improvementProgram.setUserPermissionStatus(usingImprovementProgram ? UserPermissionStatus.GRANTED : UserPermissionStatus.NOT_GRANTED);
        }
    }

    @Deprecated
    public void addTrustBadgeElementListener(TrustBadgeElementListener trustBadgeElementListener) {
        if (null != mTrustBadgeElementListeners && !mTrustBadgeElementListeners.contains(trustBadgeElementListener)) {
            mTrustBadgeElementListeners.add(trustBadgeElementListener);
        }
    }

    @Deprecated
    public void clearTrustBadgeElementListener() {
        if (null != mTrustBadgeElementListeners) {
            mTrustBadgeElementListeners.clear();
        }
    }


    @Deprecated
    public List<TrustBadgeElementListener> getTrustBadgeElementListeners() {
        return mTrustBadgeElementListeners;
    }

    public void badgeChanged(TrustBadgeElement trustBadgeElement, boolean toggled, AppCompatActivity callingActivity) {
        if (null != mBadgeListeners) {
            for (int i = 0; i < mBadgeListeners.size(); i++) {
                mBadgeListeners.get(i).onBadgeChange(trustBadgeElement, toggled, callingActivity);
            }
        }
        //Change the UserPermissionStatus as it has been switched by user
        trustBadgeElement.setUserPermissionStatus(toggled ? UserPermissionStatus.GRANTED : UserPermissionStatus.NOT_GRANTED);
    }

    public List<BadgeListener> getBadgeListeners() {
        return mBadgeListeners;
    }

    public void addBadgeListener(BadgeListener badgeListener) {
        if (null != mBadgeListeners && !mBadgeListeners.contains(badgeListener)) {
            mBadgeListeners.add(badgeListener);
        }
    }

    public void clearBadgeListener() {
        if (null != mBadgeListeners) {
            mBadgeListeners.clear();
        }
    }

    /**
     * @deprecated use rather the interface BadgeListener, called by
     * badgeChanged(TrustBadgeElement trustBadgeElement, boolean toggled, AppCompatActivity callingActivity)
     */
    @Deprecated
    public void badgeChanged(GroupType groupType, boolean toggled, AppCompatActivity callingActivity) {
        if (null != mTrustBadgeElementListeners) {
            for (int i = 0; i < mTrustBadgeElementListeners.size(); i++) {
                mTrustBadgeElementListeners.get(i).onBadgeChange(groupType, toggled, callingActivity);
            }
        }
    }


}

File: otb/src/main/java/com/orange/essentials/otb/manager/TrustBadgeManager.kt
/*
 * Copyright 2016 Orange
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* Orange Trust Badge library
 *
 * Module name: com.orange.essentials:otb
 * Version:     1.0
 * Created:     2016-03-15 by Aurore Penault, Vincent Boesch, and Giovanni Battista Accetta
 */
package com.orange.essentials.otb.manager

import android.content.Context
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import com.orange.essentials.otb.event.EventTagger
import com.orange.essentials.otb.event.EventType
import com.orange.essentials.otb.logger.Logger
import com.orange.essentials.otb.model.Term
import com.orange.essentials.otb.model.TrustBadgeElement
import com.orange.essentials.otb.model.type.ElementType
import com.orange.essentials.otb.model.type.GroupType
import com.orange.essentials.otb.model.type.UserPermissionStatus
import java.util.*

/**
 *
 * File name:   TrustBadgeManager
 * Main singleton handling the trust badge elements
 */
enum class TrustBadgeManager {
    INSTANCE;

    /**
     * mApplicationName
     * Host application name
     */
    var applicationName: String? = null
    /**
     * mApplicationPackageName
     * Host application packageName
     */
    var applicationPackageName: String? = null
        private set
    /**
     * Provides a way to tag events
     */
    //Provides a default event tagger if none has been provided
    var eventTagger: EventTagger? = null
        get() {
            if (null == field) {
                eventTagger = object : EventTagger {
                    override fun tag(eventType: EventType) {
                        Logger.d(TAG, "tagging event $eventType")
                    }

                    override fun tagElement(eventType: EventType, element: TrustBadgeElement) {
                        Logger.d(TAG, "tagging event $eventType for element $element")
                    }
                }
            }
            return field
        }
    /**
     * Store all informations about Main Data, Complementary Data, and Usage Data
     */
    var trustBadgeElements: MutableList<TrustBadgeElement> = ArrayList()
    /**
     * Stores all the listeners who should act when a toggleable badge element value is switched
     */
    val badgeListeners = ArrayList<BadgeListener>()
    /**
     * List of terms and conditions that should be displayed in "terms and conditions" section
     */
    var terms: List<Term>? = ArrayList()
    /**
     * List of Custom data fragments that should be added to the default cards and fragments already available
     */
    var customDataFragments: MutableList<CustomDataFragment> = ArrayList()
    /**
     * Boolean indicating if app checked improvement program
     */
    @Deprecated("This variable is no longer used and is kept for compatibility only. It will be removed in the near future.")
    var isUsingImprovementProgram = true

    /**
     * boolean letting us know if badges are initialized, false by default
     */
    var isInitialized = false
        private set
    /** METHODS *********************************************************************************  */
    /** KEYS DATAS  */
    /**
     * Return the host application name
     * @return Host application name from AndroidManifest.xml
     */
    private fun extractApplicationName(context: Context): String {
        var appName = context.packageName
        Logger.d(TAG, "extractApplicationName")
        val csName = context.applicationInfo.loadLabel(context.packageManager)
        if (csName.isNotEmpty()) {
            appName = csName.toString()
        }
        Logger.d(TAG, "extractApplicationName")
        return appName
    }

    /**
     * provide init with default BadgeFactory
     * @param context : Context providing needed data to retrieve permissions infos
     */
    fun initialize(context: Context) {
        PermissionManager.INSTANCE.initPermissionList(context)
        val elements = TrustBadgeElementFactory.getDefaultPermissionElements(context)
        elements.addAll(TrustBadgeElementFactory.getDefaultAppDataElements(context))
        initialize(context, elements, ArrayList())
    }

    /**
     * initialize
     * @param context            : Activity providing needed data to retrieve permissions infos
     * @param trustBadgeElements : the list of trust badge that will be used into the app
     * @param terms : the list of terms that will be used into the app
     */
    fun initialize(context: Context, trustBadgeElements: MutableList<TrustBadgeElement>, terms: List<Term>) {
        initialize(context, trustBadgeElements, terms, false)
    }

    /**
     * initialize
     * @param context            : Activity providing needed data to retrieve permissions infos
     * @param trustBadgeElements : the list of trust badge that will be used into the app
     * @param terms : the list of terms that will be used into the app
     * @param logEnabled : should log be enabled on trust badge lib : default no
     */
    fun initialize(context: Context, trustBadgeElements: MutableList<TrustBadgeElement>, terms: List<Term>, logEnabled: Boolean = false) {
        Logger.loggingAllowed = logEnabled
        Logger.d(TAG, "TrustBadgeManager initialize $trustBadgeElements")
        clean()
        PermissionManager.INSTANCE.initPermissionList(context)
        applicationName = extractApplicationName(context)
        if (!applicationName!!.isEmpty()) {
            Logger.d(TAG, "mApplicationName is NOT empty")
            applicationPackageName = context.packageName
            if (applicationPackageName != null && applicationPackageName!!.length != 0) {
                Logger.d(TAG, "mApplicationPackageName is NOT empty")
                /** Initialisation of TrustBadgeElements to 15 permissions  */
                this.trustBadgeElements = trustBadgeElements
                this.terms = terms
            } else {
                Logger.d(TAG, "Context does not provide PackageName.")
            }
        } else {
            Logger.d(TAG, "Context does not provide ApplicationName.")
        }
        isInitialized = true
    }

    /**
     * retrieve list of permissions about kids
     */
    val appDataElements: ArrayList<TrustBadgeElement>?
        get() {
            val resultPermissionList = ArrayList<TrustBadgeElement>()
            for (trustBadgeElement in this.trustBadgeElements) {
                if (trustBadgeElement.elementType == ElementType.APP_DATA) {
                    resultPermissionList.add(trustBadgeElement)
                }
            }
            return resultPermissionList
        }
    /**
     * retrieve list of  permissions
     */
    val permissionElements: ArrayList<TrustBadgeElement>?
        get() {
            val resultPermissionList = ArrayList<TrustBadgeElement>()
            for (trustBadgeElement in this.trustBadgeElements) {
                if (trustBadgeElement.elementType == ElementType.PERMISSIONS) {
                    resultPermissionList.add(trustBadgeElement)
                }
            }
            return resultPermissionList
        }

    /**
     * retrieve a specific badge according to type
     */
    fun getSpecificPermission(groupType: GroupType): TrustBadgeElement? {
        Logger.d(TAG, "getSpecificPermission")
        var result: TrustBadgeElement? = null
        for (trustBadgeElement in this.trustBadgeElements) {
            if (trustBadgeElement.groupType == groupType) {
                Logger.d(TAG, "trustBadgeElement.getElementType() equals " + groupType.toString())
                result = trustBadgeElement
            }
        }
        return result
    }
    /**
     * GETTERS & SETTERS************************************************************************
     */
    /**
     * Gets the array of trustbadge elements
     * @return the list of TrustBadgeElements that will be displayed to the user
     */
    fun addTrustBadgeElement(trustBadgeElement: TrustBadgeElement?) {
        if (null != trustBadgeElement) {
            this.trustBadgeElements.add(trustBadgeElement)
        }
    }

    /**
     * Remove all element from the SparseArray and the array
     */
    fun clearTrustBadgeElements() {
        this.trustBadgeElements.clear()
    }

    /**
     * Refresh the badges according to the permissions granted by the app
     */
    fun refreshTrustBadgePermission(context: Context) {
        Logger.d(TAG, "refreshTrustBadgePermission...")
        PermissionManager.INSTANCE.initPermissionList(context)
        for (i in this.trustBadgeElements.indices) {
            val trustBadgeElement = this.trustBadgeElements[i]
            if (trustBadgeElement.isShouldBeAutoConfigured) {
                val type = trustBadgeElement.groupType
                if (type.isSystemPermission || GroupType.NOTIFICATIONS.equals(type)) {
                    trustBadgeElement.userPermissionStatus = PermissionManager.INSTANCE.doesUserAlreadyAcceptPermission(context, trustBadgeElement.groupType)
                }
            }
            Logger.d(TAG, "refresh = $trustBadgeElement")
        }

    }

    /**
     * Method returns true if at least one badge is of type DATA or OTHER

     * @return a boolean indicating if data badges are available
     */
    fun hasPermissions(): Boolean {
        val result = permissionElements != null && permissionElements!!.isNotEmpty()
        Logger.d(TAG, "hasPermissions : $result")
        return result
    }

    /**
     * Method returns true if at least one badge is of type USAGE

     * @return a boolean indicating if usage badges are available
     */
    fun hasAppData(): Boolean {
        val result = appDataElements != null && appDataElements!!.isNotEmpty()
        Logger.d(TAG, "hasAppData : $result")
        return result
    }

    /**
     * Method returns true if at least one term is available

     * @return a boolean indicating if a term is available
     */
    fun hasTerms(): Boolean {
        val terms = terms != null && terms!!.isNotEmpty()
        Logger.d(TAG, "hasTerms : $terms")
        return terms
    }

    /**
     * Method returns true if at least one term is available

     * @return a boolean indicating if a term is available
     */
    fun hasCustomCards(): Boolean {
        return customDataFragments.size > 0
    }

    fun badgeChanged(trustBadgeElement: TrustBadgeElement, toggled: Boolean, callingActivity: AppCompatActivity) {
        for (i in this.badgeListeners.indices) {
            this.badgeListeners[i].onBadgeChange(trustBadgeElement, toggled, callingActivity)
        }
        //Change the UserPermissionStatus as it has been switched by user
        trustBadgeElement.userPermissionStatus = if (toggled) UserPermissionStatus.GRANTED else UserPermissionStatus.NOT_GRANTED
    }

    fun addBadgeListener(badgeListener: BadgeListener) {
        if (!this.badgeListeners.contains(badgeListener)) {
            this.badgeListeners.add(badgeListener)
        }
    }

    fun clearBadgeListeners() = badgeListeners.clear()

    companion object {
        /**
         * Log TAG
         */
        private const val TAG = "TrustBadgeManager"
    }

    /**
     * Add a custom fragment
     */
    fun addCustomCard(title: String, content: String?, fragment: Fragment) {
        customDataFragments.add(CustomDataFragment(title, content, fragment))
    }

    /**
     * Add a custom fragment
     */
    private fun clean() {
        customDataFragments.clear()
    }
}

data class CustomDataFragment(
        val title: String,
        val content: String? = null,
        val fragment: Fragment)
